sw/device/lib/testing/test_coverage.c

```diff
@@ -0,0 +1,75 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/log.h"
+#include "sw/device/lib/uart.h"
+#include "sw/vendor/llvm_clang_rt_profile/compiler-rt/lib/profile/InstrProfiling.h"
+#include <stdint.h>
+
+/**
+ * When the linker finds a definition of this symbol, it knows to skip loading
+ * the object which contains the profiling runtime’s static initializer. See
+ * https://clang.llvm.org/docs/SourceBasedCodeCoverage.html#using-the-profiling-runtime-without-static-initializers
+ * for more information.
+ */
+int __llvm_profile_runtime;
+
+/**
+ * Computes the CRC32 of a buffer as expected by Python's `zlib.crc32()`. The
+ * implementation below is basically a simplified, i.e. byte-by-byte and without
+ * a lookup table, version of zlib's crc32. See
+ * https://github.com/madler/zlib/blob/2fa463bacfff79181df1a5270fb67cc679a53e71/crc32.c,
+ * lines 111-112 and 276-279.
+ */
+static uint32_t crc32(uint8_t *buf, size_t len) {
+  // CRC32 polynomial.
+  static const uint32_t kPoly = 0xEDB88320;
+  // Since we use a contiguous buffer, we don't need to call this function
+  // multiple times. That's why `crc` is not a parameter and is initialized to
+  // `UINT32_MAX`, i.e. `~0`.
+  uint32_t crc = UINT32_MAX;
+  for (size_t i = 0; i < len; ++i) {
+    crc ^= buf[i];
+    for (uint8_t j = 0; j < 8; ++j) {
+      bool lsb = crc & 1;
+      crc >>= 1;
+      if (lsb) {
+        crc ^= kPoly;
+      }
+    }
+  }
+  return ~crc;
+}
+
+/**
+ * Sends the given buffer as a hex string over UART.
+ */
+static void send_buffer(uint8_t *buf, size_t len) {
+  for (uint32_t i = 0; i < len; ++i) {
+    base_printf("%2X", buf[i]);
```

Comment:
If you replace this with LOG_INFO (we can create something specific for transmitting profile data specifically if needed), then we could get the coverage collection working in DV as well.

Reply:
Thanks @sriyerg. I agree, but I prefer not to do this _right now_ because it appends a new line after each message at runtime. We plan to add a non-standard specifier to `base_printf` as @mcy suggested to print the entire buffer in a single call (please see [this thread](https://github.com/lowRISC/opentitan/pull/3080#discussion_r471634052)) and then use `LOG_INFO` to solve this issue (tracked in #3155).

Reply:
Okay, SG.